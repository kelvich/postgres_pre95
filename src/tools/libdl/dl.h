/*
 *  Ultrix 4.x Dynamic Loader Library Version 1.0
 *
 *  dl.h--
 *      header file for the Dynamic Loader Library
 *
 *
 *  Copyright (c) 1993 Andrew K. Yu, University of California at Berkeley
 *  All rights reserved.
 *
 *  Permission to use, copy, modify, and distribute this software and its
 *  documentation for educational, research, and non-profit purposes and
 *  without fee is hereby granted, provided that the above copyright
 *  notice appear in all copies and that both that copyright notice and
 *  this permission notice appear in supporting documentation. Permission
 *  to incorporate this software into commercial products can be obtained
 *  from the author. The University of California and the author make
 *  no representations about the suitability of this software for any 
 *  purpose. It is provided "as is" without express or implied warranty. 
 *
 */
#ifndef _DL_HEADER_
#define _DL_HEADER_

#include <filehdr.h>
#include <syms.h>
#include <reloc.h>
#include <scnhdr.h>
#include <fcntl.h>
#include <ar.h>

typedef long CoreAddr;


typedef struct ScnInfo {
    CoreAddr	addr;		/* starting address of the section */
    SCNHDR	hdr;		/* section header */
    RELOC      *relocEntries;	/* relocation entries */
} ScnInfo;

typedef enum {
    DL_NEEDRELOC,		/* still need relocation */
    DL_RELOCATED,		/* no relocation necessary */
    DL_INPROG			/* relocation in progress */
} dlRStatus;

typedef struct JmpTbl {
    char *block;		/* the jump table memory block */
    struct JmpTbl *next;	/* next block */
} JmpTbl;

typedef struct dlFile {
    char	*filename;	/* file name of the object file */

    int		textSize;	/* used by mprotect */
    CoreAddr 	textAddress;	/* start addr of text section */
    long     	textVaddr;	/* vaddr of text section in obj file */
    CoreAddr 	rdataAddress;	/* start addr of rdata section */
    long     	rdataVaddr;	/* vaddr of text section in obj file */
    CoreAddr 	dataAddress;	/* start addr of data section */
    long 	dataVaddr;	/* vaddr of text section in obj file */
    CoreAddr	bssAddress;	/* start addr of bss section */
    long     	bssVaddr;	/* vaddr of text section in obj file */

    int		nsect;		/* number of sections */
    ScnInfo 	*sect;		/* details of each section (array) */

    int		issExtMax;	/* size of string space */
    char 	*extss;		/* extern sym string space (in core) */
    int		iextMax;	/* maximum number of Symbols */
    pEXTR	extsyms;	/* extern syms */

    dlRStatus 	relocStatus;	/* what relocation needed? */
    int 	needReloc;

    JmpTbl	*jmptable;	/* the jump table for R_JMPADDR */

    struct dlFile *next;	/* next member of the archive */
} dlFile;

typedef struct dlSymbol {
    char *name;			/* name of the symbol */
    long addr;			/* address of the symbol */
    dlFile *objFile;		/* from which file */
} dlSymbol;

/*
 * prototypes for the dl* interface
 */
extern void *dl_open(/* char *filename, int mode */);
extern void *dl_sym(/* void *handle, char *name */);
extern void dl_close(/* void *handle */);
extern char *dl_error(/* void */);

#define   DL_LAZY	0	/* lazy resolution */
#define   DL_NOW	1	/* immediate resolution */

/*
 * Miscellaneous utility routines:
 */
extern char **dl_undefinedSymbols(/* int *count */);
extern void dl_printAllSymbols(/* void *handle */);
extern void dl_setLibraries(/* char *libs */);

#endif  _DL_HEADER_
