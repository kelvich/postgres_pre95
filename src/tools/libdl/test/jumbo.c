#include <stdio.h>
#include <math.h>
#include "dl.h"

double alpha= 2.094395;

main (argc, argv)
     int argc;
     char *argv[];
{
    void *handle;
    void (*func) ();
#if 0
    double dummy,alpha;
#endif

    (void) dl_init (argv[0]);
    dl_setLibraries("/usr/lib/libm_G0.a");

    if ((handle=dl_open("jumboobj.o", DL_NOW)) == NULL) {
	fprintf(stderr, "%s\n", dl_error());
	exit(1);
    }

    /*
     * trigo
     */
#if 0
    dummy = sin(0); dummy = cos(0); dummy = tan(0);
    dummy = fabs(1);

    printf("Enter alpha ");scanf("%lf",&alpha);
    printf("Loading file\n");
#endif

    printf("*** TRIGO ***\n");
    if (!(func= (void(*)()) dl_sym(handle, "trigo"))) {
	fprintf(stderr, "trigo: %s\n", dl_error());
    }else {
	func((double)alpha);
    }

    /*
     * testanotherbug
     */
    printf("*** TESTANOTHERBUG ***\n");
    if (!(func = dl_sym(handle, "testanotherbug"))) {
	fprintf(stderr, "testanotherbug: %s\n", dl_error());
    }else {
	(*func)(0);
    }

    /*
     * teststaticbug
     */
    printf("*** TESTSTATICBUG ***\n");
    if (!(func = dl_sym(handle, "teststaticbug"))) {
	fprintf(stderr, "teststaticbug: %s\n", dl_error());
    }else {
	(*func)(0);
    }

    /*
     * testswitchbug
     */
    printf("*** TESTSWITCHBUG ***\n");
    if (!(func = dl_sym(handle, "testswitchbug"))) {
	fprintf(stderr, "testswitchbug: %s\n", dl_error());
    }else {
	(*func)(3);
	(*func)(2);
	(*func)(1);
	(*func)(0);
    }

    printf("*** DBL ***\n");
    if (!(func = dl_sym(handle, "dbl"))) {
	fprintf(stderr, "dbl: %s\n", dl_error());
    }else {
	(*func)();
    }

    dl_close(handle);
}
