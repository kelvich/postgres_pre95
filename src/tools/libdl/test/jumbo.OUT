*** TRIGO ***
Hello
Alpha = 2.094395
sin = 0.866025 cos = -0.500000 tan = -1.732051
Goodbye
*** TESTANOTHERBUG ***
about to copy array
the string is now This is a test..
*** TESTSTATICBUG ***
about to reference ptrtyp
ptrtyp[0] = ""
ptrtyp[1] = "!"
ptrtyp[2] = "@"
ptrtyp[3] = "~"
ptrtyp[4] = "*"
ptrtyp[5] = "&"
ptrtyp[6] = "#m"
ptrtyp[7] = "#s"
ptrtyp[8] = "#p"
ptrtyp[9] = "%m"
ptrtyp[10] = "%s"
ptrtyp[11] = "%p"
ptrtyp[12] = "%"
ptrtyp[13] = "#"
ptrtyp[14] = ">"
ptrtyp[15] = ""
ptrtyp[16] = "^"
ptrtyp[17] = "\"
ptrtyp[18] = ""
ptrtyp[19] = ""
ptrtyp[20] = "+"
ptrtyp[21] = ""
ptrtyp[22] = ""
ptrtyp[23] = ""
ptrtyp[24] = ""
*** TESTSWITCHBUG ***
about to reference ptrtyp
ptrtyp[3] = "~"
about to reference ptrtyp
ptrtyp[2] = "@"
about to reference ptrtyp
ptrtyp[1] = "!"
about to reference ptrtyp
ptrtyp[0] = ""
*** DBL ***
2.0/3: 0.666667
